### ISSUE TITLE
Index of worker to parallelize tests

### ISSUE THREAD

**Issue by Fire-Dragon-DoL** [TestArmada/magellan#152]
Hi everyone, we on our team decided to use magellan.

I would like to know if there is a way to identify the worker that is running the test, for parallelization purposes.

I was planning to use `mocking_port % number_of_workers` as an identifier, but unfortunately due to port spacing, this isn't going to work (may end up with two workers having the same "index").

Are there any other way to identify which worker is running a given test? To parallelize, I need to run multiple servers at same time, so we were planning to use different urls per test server, like testerver1.localhost, testerver2.localhost, testserver3.localhost. But as you can see, I have hard times having an index of the executing worker on the test.

Any suggestion on how to achieve this?

So far, my plan is to use the PID for this purpose.

**Comment by Fire-Dragon-DoL** [comment]
Not sure where my brain was this afternoon but obviously with PID or ports this won't work. Any chance magellan keeps some kind of index that I can access to in some way, to identify workers?

**Comment by Maciek416** [comment]
Magellan's port spacing default is stable (and you can control it), so are you certain `mocking_port % number_of_workers` doesn't work? (or at least work better than `process.pid`)?

**Comment by Maciek416** [comment]
Err, you may have to modify that `%` calculation to pack it back to a mapping with no spacing.

**Comment by Fire-Dragon-DoL** [comment]
@Maciek416 you arrived at my same conclusion, since there is port spacing, the ports are not sequential, I might end up with two "3"-indexed workers. Not sure what you mean by "pack it back" (non native speaker here).

**Comment by Maciek416** [comment]
@Fire-Dragon-DoL my apologies -- I just realized (in the shower, lol) that you're totally correct. It's impossible to determine worker index from `%`, mainly because magellan is always "marching forward" in the range of available ports, and may skip ports.

Looks like we will need to add your requested feature.

**Comment by Fire-Dragon-DoL** [comment]
@Maciek416 thanks, yeah I realized the fact that I couldn't use `%` in the same way (in the shower), lol

Sorry to ask for this feature, unfortunately all the other options I have in mind are quite complex: creating a shared file and having to deal with locking and keeping a map in there or such things. Unfortunately there is no way for me to inject code inside magellan without changing magellan itself (since the settings are on a JSON file and not on a JS file), so couldn't find any real simple solution.

Thanks anyway 👍

**Comment by Fire-Dragon-DoL** [comment]
Hey @archlichking any chance this has been addressed? We are still looking for this feature

**Comment by archlichking** [comment]
@Fire-Dragon-DoL  i need your help to understand this better. through the reading i think what you need is a predictable one to one mapping to the worker (worker index) so that you can know for sure which test is assigned to which worker. however to make things much easier (and faster) magellan's worker allocator selects the idle worker randomly. the port doesn't fit your need because magellan port util will skill current port range if it is occupied. 

i don't quite understand what do you mean by `parallelism` in the first comment. can you describe your specific scenario here? we want to understand this issue better before reopen (or re-label) it.

**Comment by Fire-Dragon-DoL** [comment]
Sure @archlichking !

The idea is the following:

I have a beefy server where I run my test suite. Most of the time is lost on IO operation, waiting (the frontend itself has some operations which requires waiting 10 seconds) so basically the test suite uses no CPU and wastes time doing nothing.

Assuming my server has 8 cores, my idea is to boot 8 different backend apps (one per core) with 8 different databases. Make them accessible with 8 different url (testserver1.localhost, testserver2.localhost ... etc).

Then, I'll run Magellan which starts the test suite with 8 workers and says:

- First spawned worker, number = 1
- Second spawned worker, number = 2
- Third spawned worker, number = 3

`{ 1: Worker1Reference, 2: Worker2Reference, ... }`

 but instead of using `testserver.localhost`, it uses `testserver#{ worker_number }.localhost` as the URL.

So in the end, my e2e tests are "isolated" (they use completely different backend apps/database) and can be run in parallel.
I just need to being able to tell if the worker accessing the URL was the first spawned, or the second, or the third.
Basically is just a number from `1 to number_of_workers` assigned to each worker, so I make sure they access different backends and not all the same. I don't want a random id because makes much more complicated making up the URL (I need to create a router or put in between something like haproxy to route in the right place, just to run the test suite on different backends at same time)

Is this more clear? Having no idea how the spawning workers part is done, I'm not sure if it's possible. I was checking the codebase to try to make a pull request, but wasn't really clear to me how everything interacted

**Comment by archlichking** [comment]
@Fire-Dragon-DoL magellan doesn't assign worker id at the spawn phase. so when the test process is spawned the worker id has already been assigned. 

magellan doesn't support this right now. we'll reopen this issue.
